### PR TITLE
Fix most typehint errors

### PR DESCRIPTION
**Pull request by wsxrdv**
Some issues have not been fixed properly. See issue #13.

### CODE REVIEW

File: src/sedpack/io/dataset_filler.py [hunk 1]
```diff
@@ -153,7 +152,8 @@ def write_example(
 
         # New metadata
         previous_metadata = current_progress.shard.shard_info.custom_metadata
-        metadata_changed: bool = (custom_metadata and previous_metadata
+        metadata_changed: bool = (bool(custom_metadata)
+                                  and bool(previous_metadata)
```
Comment (jmichelp) [inline src/sedpack/io/dataset_filler.py:156]: Might be more readable using `all()`
Caveat: IIRC `all()` evaluates all expressions first whereas boolean `and` applies lazy evaluation and stops at the first item evaluated to `false`.

File: src/sedpack/io/dataset.py [hunk 2]
```diff
@@ -938,16 +959,20 @@ def as_numpy_iterator(  # pylint: disable=too-many-arguments
...
-            example_iterator = map(process_record, example_iterator)
+            example_iterator = map(process_record,
+                                   example_iterator)  # type: ignore
 
         # Randomize only if > 0 -- no shuffle in test/validation
         if shuffle:
-            example_iterator = shuffle_buffer(example_iterator,
-                                              buffer_size=shuffle)
+            example_iterator = shuffle_buffer(
+                example_iterator,  # type: ignore
```
Comment (jmichelp) [inline src/sedpack/io/dataset.py:974]: That's quite a lot of `type: ignore` lines in our code (i.e. not counting the `lz4` and other 3rd party library typing issues) at the end. Maybe have a bug to try to reduce their amount later?

File: src/sedpack/io/tfrec/read.py [hunk 3]
```diff
@@ -33,15 +33,15 @@ class IterateShardTFRec(IterateShardBase[T]):
...
 
     def __init__(
-            self,
-            dataset_structure: DatasetStructure,
-            process_record: Callable[[ExampleT], Any] | None,
-            num_parallel_calls: int = os.cpu_count(),
+        self,
+        dataset_structure: DatasetStructure,
+        process_record: Callable[[ExampleT], Any] | None,
+        num_parallel_calls: int = os.cpu_count() or 4,
```
Comment (jmichelp) [inline src/sedpack/io/tfrec/read.py:39]: Does that solves issue #8 or do we also need an upper bound?

File: tests/io/test_error.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Copyright 2023-2024 Google LLC
```
Comment (jmichelp) [inline tests/io/test_error.py:1]: Copyright 2024

The file wasn't created in 2023

File: src/sedpack/io/dataset.py (hunk 2 above)
Comment (jmichelp) [reply src/sedpack/io/dataset.py:974]: Issue #13

File: src/sedpack/io/tfrec/read.py (hunk 3 above)
Comment (wsxrdv) [reply src/sedpack/io/tfrec/read.py:39]: No. Issue #8 depends on proper benchmarking.

Comment (jmichelp) [reply src/sedpack/io/tfrec/read.py:39]: Ack

File: src/sedpack/io/dataset.py (hunk 2 above)
Comment (wsxrdv) [reply src/sedpack/io/dataset.py:974]: Issue #14

**Review by jmichelp — APPROVED** [review]

**Pull request merged**